libBlueSky/Dialog.hpp

```diff
@@ -23,29 +22,97 @@
 
 namespace BlueSky {
 
+    class SizeGrip;
+
     /**
      * @brief A Dialog
      *
      * This class automatically gives the Heaven::primaryWindow() as parent to the dialog.
      *
      */
-    class HEAVEN_BLUESKY_API Dialog : public QDialog
+    class HEAVEN_BLUESKY_API Dialog
+            : public QDialog
     {
         Q_OBJECT
+
+        class WatchDog;
+
+        friend class SheetEventFilter;
+
     public:
         Dialog();
         ~Dialog();
 
+    public:
+        bool isResizerEnabled() const;
+        void setResizerEnabled(bool enabled);
+
+    public slots:
+        virtual void open();
+
     protected:
-        void showEvent(QShowEvent*);
+#ifndef Q_OS_MAC
+        virtual void moveEvent(QMoveEvent* ev);
+        virtual void resizeEvent(QResizeEvent* ev);
+#endif
+        virtual void showEvent(QShowEvent* ev);
 
+#ifndef Q_OS_MAC
     private:
-        class WatchDog;
-        WatchDog* w;
+        bool isSheet() const;
+        void updatePinnedPos();
+#endif
+
+    private:
+        bool isSizeGripEnabled() const { return false; /* delete method */ }
+        void setSizeGripEnabled(bool enabled) const { /* delete method */ }
+
+    private:
+        WatchDog*   w;
+
+        bool        mSizeGripEnabled;
+        SizeGrip*   mResizer;
 
     private slots:
         void watchDogBark();
     };
-}
 
+#ifndef Q_OS_MAC
+    inline bool Dialog::isSheet() const
+    {
+        Qt::WindowFlags wf = windowFlags();
+        return wf.testFlag(Qt::Sheet) && !wf.testFlag(Qt::ToolTip);
+    }
+
+    inline void Dialog::updatePinnedPos()
+    {
+        QWidget* pw = parentWidget();
+        Q_ASSERT(pw);
+
+        // TODO: implement all alignment positions
+        const QRect pr = pw->geometry();
+
+        //move to top center
+        QPoint fixedPos(pr.center().x() - (width() / 2), pr.top());
+        if (pos() != fixedPos) {
+            move(fixedPos);
+        }
+    }
 #endif
+
+#ifndef Q_OS_MAC
+    class HEAVEN_BLUESKY_API SheetEventFilter : public QObject
```

Comment:
When I started implementing this, I wasn't sure about that. I guess you're right and I move it to cpp as private implementation of `BlueSky::Dialog`.